### ISSUE TITLE
fail to convert 3 channel RGB image  from python to opencv

### ISSUE THREAD

**Issue by yreme** [yati-sagade/opencv-ndarray-conversion#8]
Hi  ,  I come across 2 problems,
### 1.   rgb image could not be convert from python to opencv.

Yet when  then  image is  gray , it works:

``` python
import cv2
img=cv2.imread("img.png");

import myLib#my libs as  follows 
cv3=myLib.cv3()
cv3.show(img,"imageTitle") 
```

But  rgb image fails.
Yet, it seems that  export from opencv to python works even if  image is RGB:

``` python
imgExport=cv3.getSrcImage()
```
### 2. cv::Mat could not be reconginzed in python

error was reoport  that Mat could not  be reconginzed every time when  I call :

> cv3.srcImage

would you please give some direction on how to slove them ? Bellow is part of  CPP code.

``` cpp

class cv3 {
    public: 
    cv::Mat srcImage;

    void loadImage(const char* srcPath  ){
        srcImage =cv::imread(srcPath ); 
    } 

    void getSrcImage(){
        return srcImage ;
    }

    void show(const cv::Mat& img, std::string title ="showImage"){
        cv::imshow(title,img);
        cv::waitKey(0);
        cv::destroyWindow(title);
    } 
}   




// Wrap a few functions and classes for testing purposes
namespace fs { namespace python {


    BOOST_PYTHON_MODULE(cv3)
    {
        // Main types export
        fs::python::init_and_export_converters();
        py::scope scope = py::scope(); 



    py::class_<cv3>("cv3")   
    .add_property("srcImage", & cv3::srcImage,& cv3::srcImage,  "srcImage")
         .def("loadImage", &cv3::loadImage)
         .def("getSrcImage", &cv3::getSrcImage) 

         .def("show", &cv3::show,( py::arg("img"),py::arg("title")= std::string("showImage")  )  ) 
          ;

    }

  } // namespace python
} // namespace fs


```

**Comment by yati-sagade** [comment]
Hi. I'm sorry for the late response. Have you managed to fix this? If yes, please close this issue, maybe with a link to your solution. If not, a the full error for your second problem would be helpful :)

**Comment by yreme** [comment]
I have not to solve it yet ！ The problem is  how  to pass an rgb image  from  python to  C++ ？ Would you please give me a demo？

**Comment by yati-sagade** [comment]
If you look at the `examples.cpp` file, there is a function called `display()`. Compile the project using `make` and then put the following in a file, say, `disp.py`:

```
import examples
# Get your image here
img = ...
examples.display(img)
```

Running this file should display any image, RGB or not. It works for me. Let me know what errors you get if it doesn't work for you.

**Comment by yreme** [comment]
> make

and the result is

```
g++ -g -Wall -std=c++11 -I/usr/local/include $(pkg-config --cflags opencv) -I/opt/cuda/include $(pkg-config --cflags python2) \
    -fPIC -c examples.cpp
In file included from /usr/include/python2.7/numpy/ndarraytypes.h:1761:0,
                 from /usr/include/python2.7/numpy/ndarrayobject.h:17,
                 from conversion.h:8,
                 from examples.cpp:5:
/usr/include/python2.7/numpy/npy_1_7_deprecated_api.h:15:2: warning: #warning "Using deprecated NumPy API, disable it by " "#defining NPY_NO_DEPRECATED_API NPY_1_7_API_VERSION" [-Wcpp]
 #warning "Using deprecated NumPy API, disable it by " \
  ^
examples.cpp: In function ‘PyObject* mul(PyObject*, PyObject*)’:
examples.cpp:57:10: warning: unused variable ‘r1’ [-Wunused-variable]
     auto r1 = leftMat.rows, c1 = leftMat.cols, r2 = rightMat.rows,
          ^
examples.cpp:58:10: warning: unused variable ‘c2’ [-Wunused-variable]
          c2 = rightMat.cols;
          ^
In file included from examples.cpp:5:0:
conversion.h: At global scope:
conversion.h:13:18: warning: ‘opencv_error’ defined but not used [-Wunused-variable]
 static PyObject* opencv_error = 0;
                  ^
conversion.h:15:12: warning: ‘int failmsg(const char*, ...)’ declared ‘static’ but never defined [-Wunused-function]
 static int failmsg(const char *fmt, ...);
            ^
conversion.h:33:18: warning: ‘PyObject* failmsgp(const char*, ...)’ declared ‘static’ but never defined [-Wunused-function]
 static PyObject* failmsgp(const char *fmt, ...);
                  ^
g++ -g -Wall -std=c++11 -I/usr/local/include $(pkg-config --cflags opencv) -I/opt/cuda/include $(pkg-config --cflags python2) \
    -fPIC -c conversion.cpp
In file included from /usr/include/python2.7/numpy/ndarraytypes.h:1761:0,
                 from /usr/include/python2.7/numpy/ndarrayobject.h:17,
                 from conversion.h:8,
                 from conversion.cpp:1:
/usr/include/python2.7/numpy/npy_1_7_deprecated_api.h:15:2: warning: #warning "Using deprecated NumPy API, disable it by " "#defining NPY_NO_DEPRECATED_API NPY_1_7_API_VERSION" [-Wcpp]
 #warning "Using deprecated NumPy API, disable it by " \
  ^
In file included from conversion.cpp:1:0:
conversion.h:13:18: warning: ‘opencv_error’ defined but not used [-Wunused-variable]
 static PyObject* opencv_error = 0;
                  ^
conversion.cpp:51:18: warning: ‘PyObject* failmsgp(const char*, ...)’ defined but not used [-Wunused-function]
 static PyObject* failmsgp(const char *fmt, ...)
                  ^
conversion.cpp:7:13: warning: ‘void init()’ defined but not used [-Wunused-function]
 static void init()
             ^
g++ -shared  -Wl,--export-dynamic -L/usr/local/lib \
    -L/opt/cuda/lib$(getconf LONG_BIT) $(pkg-config --libs opencv) -lboost_python $(pkg-config --libs python2) \
    examples.o conversion.o -o examples.so
```

But when  I try to import   `examples.so`, error  reported:

> > > import examples
> > > Traceback (most recent call last):
> > >   File "<stdin>", line 1, in <module>
> > > ImportError: ./examples.so: undefined symbol: _ZNK5boost6python7objects21py_function_impl_base9max_arityEv

**Comment by yati-sagade** [comment]
This typically means there's some problem linking with your boost installation. @ppalasek had the same issue and has fixed it. See issue #6

**Comment by yreme** [comment]
Thanks for your code, I complied it by cmake.Now rgb image could pass from python to opencv (^o^)/~.

## Build the files

cd  the root of files

```
mkdir build
cd build 
cmake ..
make
```

## pass RGB image  from python to opencv 

here is an example that works

``` python

import os
# path of examples.so
path_of_so_file="/home/emery/opensource/opencv/demos/opencv-ndarray-conversion-fix/build"
os.chdir(path_of_so_file)
import examples


import cv2

# Get your image here
imgPath = '/home/emery/Pictures/Pic/1.jpg'


img=cv2.imread(imgPath);
examples.display(img) 

```

## demo code

since github could not upload zip files. one could download this image and rename it as `opencv-ndarray-conversion.zip` , and get the codes. ![opencv-ndarray-conversion-f zip](https://cloud.githubusercontent.com/assets/6118517/5350234/c478ec58-7f7c-11e4-9893-2024a70154a4.png)

**Comment by yati-sagade** [comment]
That's great! If you'd like to make a pull request with a proper `CMakeList.txt` and docs, I'd be happy to merge it. See issue #7. Closing this one.

**Comment by yreme** [comment]
I do not plan to make  a pull request. But if you  merge this, it will 
be very nice O(∩_∩)O！

**Comment by JohnnyY8** [comment]
Hi @yreme : 
         I saw this issue that you can convert 3 channel RGB image from python (ndarray) to opencv (MAT). I want to do the same thing, but still can't finish it. According to the README, I have fixed some bug following the #6 and managed to run make and make test. Yet, I still don't know how to use these functions in my own codes.
        Could you please give me any more help?
        Thanks a lot!
